core/unix/os.c

```diff
@@ -2518,8 +2518,9 @@ os_tls_calloc(DR_PARAM_OUT uint *offset, uint num_slots, uint alignment)
     d_r_mutex_lock(&client_tls_lock);
     for (i = 0; i < MAX_NUM_CLIENT_TLS; i++) {
         if (!client_tls_allocated[i] &&
-            /* ALIGNED doesn't work for 0 */
-            (alignment == 0 || ALIGNED(offs + i * sizeof(void *), alignment))) {
+            (start != -1 ||
+             /* ALIGNED doesn't work for 0 */
+             alignment == 0 || ALIGNED(offs + i * sizeof(void *), alignment))) {
```

Comment:
It would be great to add a test of this.  It needs too much setup to easily mock so it can't join the unit tests at the bottom of this file, so it would have to be inside a client test: please add to suite/tests/client-interface/tls.dll.cpp which should be straightforward: call it in a loop with free calls.  That test should check the return value too.

Reply:
I've added a test, but it fails on some Windows targets. The test assumes that allocation comes from a single contiguous block. Is that perhaps not true on some Windows? I could make the test Unix-only.

Reply:
The Windows allocator seems to be here: https://github.com/DynamoRIO/dynamorio/blob/7f7544b6f593dcdbb2213fca3922055092b35af1/core/win32/ntdll.c#L1684

It's quite complicated, and the word `FIXME` appears in about six places.

Reply:
`FIXME` was used in early days for what would be `XXX` today and so does not indicate by itself anything alarming.

Reply:
In general, Windows has to steal slots from the application, and so it can be very constrained, unlike UNIX which has its own dedicated mapping.

Reply:
Looks like this test only allocates 4 slots at once: that would be expected to work for the test app being run which does not grab all the slots itself.  What is the test failing on: which assert?

Reply:
On Windows I am now freeing each allocation separately and exactly once.